node/src/components/consensus/protocols/highway/synchronizer.rs

```diff
@@ -0,0 +1,254 @@
+
+/// An incoming vertex that we haven't added to the protocol state yet.
+#[derive(DataSize, Debug)]
+pub(crate) struct PendingVertex<I, C>
+where
+    C: Context,
+{
+    /// The peer who sent it to us.
+    sender: I,
+    /// The pre-validated vertex.
+    pvv: PreValidatedVertex<C>,
+    /// The time when we received it.
+    time_received: Timestamp,
+}
+
+impl<I, C: Context> PendingVertex<I, C> {
+    /// Returns a new pending vertex with the current timestamp.
+    pub(crate) fn new(sender: I, pvv: PreValidatedVertex<C>) -> Self {
+        Self {
+            sender,
+            pvv,
+            time_received: Timestamp::now(),
+        }
+    }
+
+    pub(crate) fn sender(&self) -> &I {
+        &self.sender
+    }
+
+    /// Returns the vertex waiting to be added.
+    pub(crate) fn vertex(&self) -> &Vertex<C> {
+        self.pvv.inner()
+    }
+
+    pub(crate) fn pvv(&self) -> &PreValidatedVertex<C> {
+        &self.pvv
+    }
+
+    /// Returns `true` if this vertex was received longer than `timeout` ago.
+    fn expired(&self, timeout: TimeDiff) -> bool {
+        self.time_received + timeout < Timestamp::now()
+    }
+}
+
+impl<I, C: Context> Into<PreValidatedVertex<C>> for PendingVertex<I, C> {
+    fn into(self) -> PreValidatedVertex<C> {
+        self.pvv
+    }
+}
+
+#[derive(DataSize, Debug)]
+pub(crate) struct Synchronizer<I, C>
+where
+    C: Context,
+{
+    /// Incoming vertices we can't add yet because they are still missing a dependency.
+    vertex_deps: BTreeMap<Dependency<C>, Vec<PendingVertex<I, C>>>,
+    /// The vertices that are scheduled to be processed at a later time.  The keys of this
+    /// `BTreeMap` are timestamps when the corresponding vector of vertices will be added.
+    vertices_to_be_added_later: BTreeMap<Timestamp, Vec<PendingVertex<I, C>>>,
+    /// Vertices that might be ready to add to the protocol state: We are not currently waiting for
+    /// a requested dependency.
+    vertices_to_be_added: Vec<PendingVertex<I, C>>,
+    /// The duration for which incoming vertices with missing dependencies are kept in a queue.
+    pending_vertex_timeout: TimeDiff,
+}
+
+impl<I: NodeIdT, C: Context + 'static> Synchronizer<I, C> {
+    pub(crate) fn new(pending_vertex_timeout: TimeDiff) -> Self {
+        Synchronizer {
+            vertex_deps: BTreeMap::new(),
+            vertices_to_be_added_later: BTreeMap::new(),
+            vertices_to_be_added: Vec::new(),
+            pending_vertex_timeout,
+        }
+    }
+
+    /// Removes expired pending vertices from the queues, and schedules the next purge.
+    pub(crate) fn purge_vertices(&mut self) {
+        let timeout = self.pending_vertex_timeout;
+        self.vertices_to_be_added.retain(|pv| !pv.expired(timeout));
+        for pvs in self
+            .vertices_to_be_added_later
+            .values_mut()
+            .chain(self.vertex_deps.values_mut())
+        {
+            pvs.retain(|pv| !pv.expired(timeout));
+        }
```

Comment:
What do you think about a method that removes expired entries from a collection of `PendingVertex`? It would replace those two and will reduce `purge_vertices` to:
```rust
Self::remove_expired(self.vertices_to_be_added);
self
  .vertices_to_be_added_later
  .values_mut()
  .chain(self.vertex_deps.values_mut())
  .for_each(Self::remove_expired)
```

Reply:
Doesn't work quite like that; I think the most readable solution might be to just extract the whole logic of removing expired entries _and_ then removing empty lists from _maps_. I'll give that a try.